Habitica/src/main/java/com/habitrpg/android/habitica/data/implementation/ChallengeRepositoryImpl.java

```diff
@@ -0,0 +1,123 @@
+import com.habitrpg.android.habitica.models.social.Challenge;
+import com.habitrpg.android.habitica.models.social.PostChallenge;
+import com.habitrpg.android.habitica.models.tasks.Task;
+import com.habitrpg.android.habitica.models.tasks.TaskList;
+import com.habitrpg.android.habitica.models.tasks.TasksOrder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import rx.Observable;
+import rx.schedulers.Schedulers;
+
+
+public class ChallengeRepositoryImpl extends BaseRepositoryImpl<ChallengeLocalRepository> implements ChallengeRepository {
+
+    public ChallengeRepositoryImpl(ChallengeLocalRepository localRepository, ApiClient apiClient) {
+        super(localRepository, apiClient);
+    }
+
+    @Override
+    public Observable<Challenge> getChallenge(String challengeId) {
+        return apiClient.getChallenge(challengeId);
+    }
+
+    @Override
+    public Observable<TaskList> getChallengeTasks(String challengeId) {
+        return apiClient.getChallengeTasks(challengeId);
+    }
+
+
+    private TasksOrder getTaskOrders(List<Task> taskList) {
+        Map<String, List<Task>> stringListMap = $.groupBy(taskList, t -> t.getType());
+
+        TasksOrder tasksOrder = new TasksOrder();
+
+        for (Map.Entry<String, List<Task>> entry : stringListMap.entrySet()) {
+            List<String> taskIdList = $.map(entry.getValue(), t -> t.getId());
+
+            switch (entry.getKey()) {
+                case Task.TYPE_HABIT:
+                    tasksOrder.setHabits(taskIdList);
+                    break;
+                case Task.TYPE_DAILY:
+                    tasksOrder.setDailys(taskIdList);
+                    break;
+                case Task.TYPE_TODO:
+                    tasksOrder.setTodos(taskIdList);
+                    break;
+                case Task.TYPE_REWARD:
+                    tasksOrder.setRewards(taskIdList);
+                    break;
+            }
+        }
+
+        return tasksOrder;
+    }
+
+    private Observable addChallengeTasks(String challengeId, List<Task> addedTaskList) {
+        if (addedTaskList.size() == 1) {
+            return apiClient.createChallengeTask(challengeId, addedTaskList.get(0));
+        } else {
+            return apiClient.createChallengeTasks(challengeId, addedTaskList);
+        }
+    }
+
+    @Override
+    public Observable<Challenge> createChallenge(PostChallenge challenge, List<Task> taskList) {
+        challenge.tasksOrder = getTaskOrders(taskList);
+
+        return Observable.create(subscriber -> {
+            apiClient.createChallenge(challenge).subscribe(challenge1 -> {
+                addChallengeTasks(challenge1.id, taskList).subscribe(task -> {
+                    subscriber.onNext(challenge1);
+                    subscriber.onCompleted();
+                }, throwable -> subscriber.onError((Throwable) throwable));
+
+            }, throwable -> subscriber.onError(throwable));
+        });
+    }
```

Comment:
can't find any operator, many of them want to have the same type or just not await the prior Observable.

for example using .map / .flatMap on `apiClient.createChallenge(..)` with `addChallengeTasks`  I'd receive `Observable<Observable>` which is also not a working solution to try `.zip` if this would work with different types

@TheHollidayInn do you have any idea ?